ITE/10/README.adoc

```diff
@@ -0,0 +1,265 @@
+  (recorded as the subject of such attestations) and produce attributes about
+  them as the outcome. Eg. test results, reviews.
+
+Note that predicates that fit the first category may also record certain
+attributes. For example, the environment field in in-toto links can include such
+attribute information.
+
+With this representation of different predicate types, step declarations in
+layouts can _chain_ different operations represented by distinct predicate types
+together, a fundamental feature of in-toto. For example, a code review step
+represented by a review predicate type that follows a checkout step that fetches
+source code represented by an in-toto link can use in-toto's artifact rules to
+ensure the review was performed against the sources fetched by the checkout
+step.
+
+=== Constraints
+
+As noted before, current in-toto layouts constraints are expressed via artifact
+rules. As the previous in-toto verification model primarily focused on steps
+that affected artifacts, constraints on custom attributes recorded in opaque
+fields such as environment and byproducts were left to inspections. With the
+rise of attribute focused predicates, in-toto layouts must support setting
+constraints on their contents.
+
+As such, this ITE proposes the following schema for step declarations.
+
+```yaml
+name: string
+command: string
+predicates: list of TypeURIs
+expectedMaterials: list of constraints
+expectedProducts: list of constraints
+expectedAttributes: list of constraints
+functionaries: list of strings
+threshold: int
+```
+
+Each entry in `expectedMaterials`, `expectedProducts`, and `expectedAttributes`
+corresponds to one rule, similar to the current in-toto layouts. However, along
+with in-toto artifact rules, other DSLs may be used to specify attribute
+specific constraints. Languages such as Rego, CUE, and CEL have been designed to
+enable such schema and data validation. Each constraint must identify the DSL
+used to ensure the verification engine can appropriately handle the rules. In
+some cases, the actual rules may need to be encoded in some form prior to
+embedding in the layout. The verification engine uses the information about the
+language to identify how to parse the rule.
+
+```yaml
+language: string
+rule: object
+```
+
+=== Verification Workflow
+
+After verifying the layout and attestation signatures (including thresholds
+where applicable), attestations must be parsed to identify the materials and
+products of the step they represent. In link type predicates, the materials are
+found within the predicate and products are recorded as the attestation's
+subject. Note that alongside these artifacts, other predicate fields excluding
+the materials must also be parsed in as part of the products to enable policies
+on their contents.  For attribute type predicates, the materials are the
+attestation's subject and products are exclusively the attributes recorded
+within the predicate.
```

Comment:
I feel like something is missing here. A couple questions: How do verifiers map predicate types in the `predicates` field to whether they are link class or attribute class? Also, how is the `expectedAttributes` used? This description only talks about how materials and product rules may be applied in the attribute class case.

Reply:
Also added some details to the verification workflow though I think it's not going to yet answer everything.

Reply:
> As with the current in-toto specification, the verification workflow is
presented with a root layout, public keys to verify the layout's signatures, and
a bundle of in-toto attestations. As the verifier expects a certain set of
predicate types, it is also expected to be aware of the class of each predicate.

Is this a reasonable expectation?